File: study/1주차 문제/최혜령/올바른괄호.py
def solution(s):
    stack = []
    i = 0
    while i < len(s):
        if len(stack) > 1:
            if stack[-1] == ')' and stack[-2] == '(':
                stack.pop(-1)
                stack.pop(-1)
            else:
                stack.append(s[i])
                i += 1
        else:
            stack.append(s[i])
            i += 1
    if len(stack) == 2 and stack[0] == '(' and stack[1] == ')': return True

    return False

File: study/1주차 문제/이다영/올바른 괄호.java
package fin;

import java.util.Stack;

public class Pro_12909 {
	public static void main(String[] args) {
		System.out.println(solution("()()"));
		System.out.println(solution("(())()"));
		System.out.println(solution(")()("));
		System.out.println(solution("(()("));
	}
	
	public static boolean solution(String s) {
		
		Stack<String> stack = new Stack<>();
		
		for (int i = 0; i < s.length(); i++) {
			String tmp = s.charAt(i) + "";
			
			if (tmp.equals("(")) stack.push(tmp);
			else {
				if (stack.size() == 0) return false;
				else {
					String peek = stack.pop();
					
					if (peek.equals(")")) return false;
				}
			}
			
			if (stack.size() > s.length()-(i+1)) return false;
		}
		
		return true;
	}
}
